globus_action_provider_tools/authentication.py

```diff
@@ -152,42 +146,113 @@ def dependent_tokens_cache_id(self) -> Optional[str]:
+        and looking for the requested scope in the set of tokens returned by the
+        grant. If a refresh token is present in the grant response, a
+        RefreshTokenAuthorizer is returned. If no refresh token is present, but an access
+        token is present, an AccessTokenAuthorizer will be returned.
+
+        A returned AccessTokenAuthorizer is guaranteed to be usable for at least the
+        value passed in via required_authorizer_expiration_time which defaults to 60
+        seconds. This implies that the authorizer, and the client created from it will be
+        usable for at least this long. It is possible that the access token in the
+        authorizer may expire after a time greater than this limit.
+
+        If no dependent tokens can be generated for the requested scope, a None value is
+        returned.
+
+        To avoid redundant calls to perform dependent token grants, the class
+        caches dependent token results for a particular incoming Bearer token used to
+        access this Action Provider.
+
+        If for any reason the caller of this function does not want to use a cached
+        result, but would require that a new dependent grant is performed, the
+        bypass_dependent_token_cache parameter may be set to True. This is used
+        automatically in a case where an access token retrieved from cache is already
+        expired: the function is called recursively to perform a new dependent token
+        grant to get a new, valid token (even though bypass is set, the new token value
+        will be added to the cache).
+
+        """
         try:
-            dep_tkn_resp = self.get_dependent_tokens().by_scopes[scope]
+            dep_tkn_resp = self.get_dependent_tokens(
+                bypass_cache_lookup=bypass_dependent_token_cache
+            ).by_scopes[scope]
         except KeyError:
             return None
 
-        if "refresh_token" in dep_tkn_resp:
-            return RefreshTokenAuthorizer(
-                dep_tkn_resp["refresh_token"],
-                self.auth_client,
-                access_token=dep_tkn_resp["access_token"],
-                expires_at=dep_tkn_resp["expires_at_seconds"],
-            )
-        elif "access_token" in dep_tkn_resp:
-            return AccessTokenAuthorizer(dep_tkn_resp["access_token"])
-        else:
-            return None
+        refresh_token = dep_tkn_resp.get("refresh_token")
+        access_token = dep_tkn_resp.get("access_token")
+        token_expiration = dep_tkn_resp.get("expires_at_seconds", 0)
+        # IF for some reason the token_expiration comes in a string, or even a string
+        # containing a float representation, try converting to a proper int. If the
+        # conversion is impossible, set expiration to 0 which should force some sort of
+        # refresh as described elsewhere.
+        if not isinstance(token_expiration, int):
+            try:
+                token_expiration = int(float(token_expiration))
```

Comment:
A bit of trivia (that will never happen within the context of our services): it's possible for this code to raise an `OverflowError`, too!

```python
int(float("inf"))  # 💥 OverflowError: cannot convert float infinity to integer
```

Within the context of our API's this will never happen but it's possible when interacting with untrusted sources.